File: Codes in C++/Array/Koko Eating Bananas.cpp
class Solution {
public:
    long totalHours(vector<int>& v, int hourly) {
        long totalH = 0;
        for (int i = 0; i < v.size(); i++) {
            totalH += ceil((double)v[i] / (double)hourly);
        }
        return totalH;
    }
    int minEatingSpeed(vector<int>& piles, int h) {
        int low = 1, high = *max_element(piles.begin(), piles.end());
        while (low <= high) {
            int mid = (low + high) / 2;
            long totalH = totalHours(piles, mid);
            if (totalH <= h) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }
};

File: Codes in Python3/Binary Search/Koko Eating Bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l=1
        r=max(piles)
        res=r

        while l<=r:
            k=(l+r)//2
            totalTime=0
            for p in piles:
                totalTime+=math.ceil(float(p)/float(k))
            if totalTime<=h:
                res=k
                r=k-1
            else:
                l=k+1
        return res

# TC=O(nlogm)
# SC=O(1)